File: compiler/nodes/statements/constant_definition.go
package statements

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/attributes"
	"wipple/nodes/constraints"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type MissingConstantValueFact struct{}

func (fact MissingConstantValueFact) String() string {
	return "is missing constant value"
}

type ConstantDefinitionNode struct {
	Comments    []string
	Attributes  []*attributes.AttributeNode
	Name        string
	Type        database.Node
	Constraints []database.Node
	Facts       *database.Facts
}

func (node *ConstantDefinitionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseConstantDefinitionStatement(parser *syntax.Parser) (*ConstantDefinitionNode, *syntax.Error) {
	span := parser.Spanned()

	comments, err := ParseComments(parser)
	if err != nil {
		return nil, err
	}

	attributes, err := attributes.ParseAttributes(parser)
	if err != nil {
		return nil, err
	}

	name, err := syntax.ParseVariableName(parser)
	if err != nil {
		return nil, err
	}

	ty, constraints, err := ParseConstantConstraints(parser)
	if err != nil {
		return nil, err
	}

	return &ConstantDefinitionNode{
		Comments:    comments,
		Attributes:  attributes,
		Name:        name,
		Type:        ty,
		Constraints: constraints,
		Facts:       database.NewFacts(span()),
	}, nil
}

func ParseConstantConstraints(parser *syntax.Parser) (database.Node, []database.Node, *syntax.Error) {
	_, err := parser.Token("AnnotateOperator", syntax.TokenConfig{
		Commit: "in this constant definition",
	})
	if err != nil {
		return nil, nil, err
	}

	parser.ConsumeLineBreaks()

	constantType, err := types.ParseType(parser)
	if err != nil {
		return nil, nil, err
	}

	constraints, _, err := syntax.ParseOptional(parser, constraints.ParseConstraints)
	if err != nil {
		return nil, nil, err
	}

	return constantType, constraints, nil
}

func (node *ConstantDefinitionNode) Visit(visitor *visit.Visitor) {
	visit.Defining(visitor, node, func() (*visit.ConstantDefinition, bool) {
		visitor.PushScope()

		visitor.AfterTypeDefinitions(func() {
			visitor.CurrentDefinition.WithImplicitTypeParameters(func() {
				visitor.Visit(node.Type)

				for _, constraint := range node.Constraints {
					visitor.Visit(constraint)
				}
			})

			visitor.Constraint(typecheck.GroupConstraint(node, node.Type))
		})

		visitor.PopScope()

		definition := &visit.ConstantDefinition{
			Name:       node.Name,
			Node:       node,
			Comments:   node.Comments,
			Attributes: visit.ParseConstantAttributes(node.Attributes),
			Assigned:   false,
			Value:      node.Type,
		}

		visitor.Define(node.Name, definition)

		visitor.AfterAllExpressions(func() {
			if !definition.Assigned {
				database.SetFact(node, MissingConstantValueFact{})
			}
		})

		return definition, true
	})
}

func (node *ConstantDefinitionNode) Codegen(c *codegen.Codegen) error {
	// Handled specially in `codegen.Codegen`
	return nil
}

File: compiler/src/nodes/statements/constant_definition.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Db, Fact, Node, NodeRef, Render},
    nodes::{parse_attributes, parse_comments, parse_constraints, parse_type},
    syntax::{ParseError, Parser, TokenKind, parse_variable_name},
    typecheck::GroupConstraint,
    visit::{ConstantAttributes, ConstantDefinition, Defined, Definition, Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct MissingConstantValue;

impl Fact for MissingConstantValue {}

impl Render for MissingConstantValue {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "is missing constant value")
    }
}

#[derive(Debug)]
pub struct ConstantDefinitionNode {
    pub comments: Vec<String>,
    pub attributes: Vec<NodeRef>,
    pub name: String,
    pub ty: NodeRef,
    pub constraints: Vec<NodeRef>,
}

impl Node for ConstantDefinitionNode {}

pub fn parse_constant_definition_statement(
    parser: &mut Parser<'_>,
) -> Result<ConstantDefinitionNode, ParseError> {
    let comments = parse_comments(parser)?;
    let attributes = parse_attributes(parser)?;
    let name = parse_variable_name(parser)?;
    let (ty, constraints) = parse_constant_constraints(parser)?;

    Ok(ConstantDefinitionNode {
        comments,
        attributes,
        name,
        ty,
        constraints,
    })
}

pub fn parse_constant_constraints(
    parser: &mut Parser<'_>,
) -> Result<(NodeRef, Vec<NodeRef>), ParseError> {
    parser.token(TokenKind::AnnotateOperator)?;
    parser.commit("in this constant definition");
    parser.consume_line_breaks();

    let ty = parse_type(parser)?;

    let constraints = parser
        .parse_optional(parse_constraints)?
        .unwrap_or_default();

    Ok((ty, constraints))
}

impl Visit for ConstantDefinitionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visitor.defining(node, |visitor| {
            visitor.push_scope();

            visitor.after_type_definitions({
                let ty = self.ty.clone();
                let constraints = self.constraints.clone();
                let node = node.clone();
                move |visitor| {
                    visitor.with_implicit_type_parameters(|visitor| {
                        visitor.visit(&ty);
                        visitor.edge(&ty, &node, "type");

                        for constraint in &constraints {
                            visitor.visit(constraint);
                            visitor.edge(constraint, &node, "constraint");
                        }
                    });

                    visitor.constraint(GroupConstraint::new(node, ty));
                }
            });

            visitor.pop_scope();

            visitor.edge(&self.ty, node, "type");
            visitor.constraint(GroupConstraint::new(node.clone(), self.ty.clone()));

            let definition = ConstantDefinition {
                name: self.name.clone(),
                node: node.clone(),
                comments: self.comments.clone(),
                attributes: ConstantAttributes::from_attributes(visitor.db, &self.attributes),
                value: None,
            };

            visitor.define(&definition.name, definition.clone());

            let node = node.clone();
            visitor.after_all_expressions(move |visitor| {
                let Some(Defined(definition)) = visitor.db.get(&node) else {
                    return;
                };

                if let Definition::Constant(definition) = definition
                    && definition.value.is_none()
                {
                    visitor.db.insert(&node, MissingConstantValue);
                }
            });

            definition
        });
    }
}

impl Codegen for ConstantDefinitionNode {
    fn codegen(&self, _codegen: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Ok(())
    }
}
